snoop/data/digests.py

```diff
@@ -142,15 +130,40 @@ def gather(blob, **depends_on):
             rv['location'] = exif_data.get('location')
             rv['date-created'] = exif_data.get('date-created')
 
+    digest, _ = models.Digest.objects.get_or_create(blob=blob)
+    rv = defaultdict([], rv)
+
+    # Entity Extraction and language detection
+    if settings.EXTRACT_ENTITIES:
+        text = rv.get('text', '')
+        if text:
+            rv.update(entities.extract_enitities(text, 'text', digest))
+        if ocr_results:
+            for ocr_name, ocrtext in rv.get('ocrtext'):
+                if ocrtext:
+                    ocr_results = entities.extract_enitities(ocrtext, ocr_name, digest)
+                    if 'entities' in rv:
+                        rv['entities'].extend(ocr_results['entities'])
+                        rv['ent-ids'].extend(ocr_results['ent-ids'])
+                        for k, v in ocr_results:
+                            if k.startswith('entity-type.'):
+                                rv[k].extend(v)
+                    else:
+                        rv.update(ocr_results)
+    elif settings.DETECT_LANGUAGE:
+        text = rv.get('text', '')[:2500]
+        if text:
+            rv.update(entities.get_language(text))
+        if ocr_results:
+            for ocr_name, ocrtext in rv.get('ocrtext'):
+                if ocrtext:
+                    rv[f'{ocr_name}_lan'] = entities.get_language(ocrtext[:2500])
+
     with models.Blob.create() as writer:
         writer.write(json.dumps(rv).encode('utf-8'))
 
-    _, _ = models.Digest.objects.update_or_create(
-        blob=blob,
-        defaults=dict(
-            result=writer.blob,
-        ),
-    )
+        digest.result = writer.blob
```

Comment:
this needs to be outside the `as writer` block, see the original code: 

```python
with models.Blob.create() as writer:
    writer.write(json.dumps(rv).encode('utf-8'))

# rest of the stuff on this same level here, not in block above:
...
```

The "writer" is a smart context manager that only creates the `.blob` on exiting the context.